content/blog/tidymodels-2021-q2/index.Rmd

```diff
@@ -0,0 +1,232 @@
+- [tidyposterior](https://github.com/tidymodels/tidyposterior/blob/master/NEWS.md#tidyposterior-010)
+- [tune](https://tune.tidymodels.org/news/index.html#tune-0-1-5-2021-04-23)
+- [workflowsets](https://workflowsets.tidymodels.org/news/index.html#workflowsets-0-0-2-2021-04-16)
+- [yardstick](https://yardstick.tidymodels.org/news/index.html#yardstick-0-0-8-2021-03-28)
+
+
+## Reduce the memory footprint of your recipes
+
+The [butcher](https://butcher.tidymodels.org/) package provides methods to remove (or "axe") components from model objects that are not needed for prediction. There are many methods for various kinds of models, but the most recent release updated how butcher handled _recipes_, the tidymodels' approach for preprocessing and feature engineering, for more complete and robust coverage. Let's consider a simulated churn classification data set for phone company customers:
+
+```{r, fig.width=8, fig.height=5, message=FALSE}
+library(tidymodels)
+library(butcher)
+data("mlc_churn")
+
+set.seed(123)
+churn_split <- initial_split(mlc_churn)
+churn_train <- training(churn_split)
+churn_test  <- testing(churn_split)
+
+ggplot(churn_train, aes(y = voice_mail_plan, fill = churn)) +
+  geom_bar(alpha = 0.8, position = "fill") +
+  scale_x_continuous(labels = scales::percent_format()) +
+  labs(x = NULL)
+```
+
+For some kinds of models, we would want to create dummy or indicator variables from nominal predictors and preprocess features to be on the same scale. We can use recipes for this task:
+
+```{r}
+churn_rec <- 
+  recipe(churn ~ voice_mail_plan + total_intl_minutes + 
+           total_day_minutes + total_eve_minutes + state, 
+         data = churn_train) %>%
+  step_dummy(all_nominal_predictors()) %>%
+  step_normalize(all_predictors())
+```
+
+You can `prep(churn_rec)` to estimate the quantities needed to create categorical features and to scale all the predictors:
+
+```{r}
+churn_prep <- prep(churn_rec)
+churn_prep
+```
+
+To remove everything from this prepped recipe not needed for applying to new data (e.g. [bake()](https://recipes.tidymodels.org/reference/bake.html) it), we can call `butcher(churn_prep)`. In some applications, modeling practitioners need to make custom functions with a feature engineering recipe. Sometimes those functions have... a lot of extra STUFF in them, stuff that is not needed for prediction.
+
+```{r}
+butchered_plus <- function() {
+  some_junk_in_the_environment <- runif(1e6)
+  
+  churn_prep <- 
+    recipe(churn ~ voice_mail_plan + total_intl_minutes + 
+             total_day_minutes + total_eve_minutes + state, 
+           data = churn_train) %>%
+    step_dummy(all_nominal_predictors()) %>%
+    step_normalize(all_predictors()) %>%
+    prep()
+  
+  butcher(churn_prep)
+}
+```
+
+In the old version of butcher, we did not successfully remove all that extra stuff, and recipes were bigger than they needed to be:
+
+```{r, eval=FALSE}
+lobstr::obj_size(butcher(churn_prep))
```

Comment:
I might add a comment above the code here giving the version number of butcher used, or at least say `# old version of butcher` and `# new version of butcher` (or something to that effect). My theory is that some readers pretty much look at the code exclusively, and might not catch the difference (esp since it's shown via `eval = FALSE`/not something they'll reproduce by running the code themselves).